multicluster/controllers/multicluster/leader/member_cleanup_controller.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2023 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leader
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	mcv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+)
+
+const indexKey = "spec.clusterID"
+
+var (
+	getResourceExportsByClusterIDFunc = getResourceExportsByClusterID
+
+	// cleanUpRetry is the retry when the cleanUpStaleResourceExports
+	// failed to clean up all stale ResourceExports.
+	cleanUpRetry = wait.Backoff{
+		Steps:    5,
+		Duration: 500 * time.Millisecond,
+		Factor:   1.0,
+		Jitter:   1,
+	}
+
+	ignoreUpdateEventPredicate = predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+	}
+)
+
+// MemberResourcesCleanupController will remove all ResourceExports belong to a member
+// cluster when the corresponding MemberClusterAnnounce CR is deleted. It will also try
+// to clean up all stale ResourceExports during start.
+type MemberResourcesCleanupController struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	namespace string
+}
+
+func NewMemberResourcesCleanupController(Client client.Client,
+	Scheme *runtime.Scheme,
+	namespace string) *MemberResourcesCleanupController {
+	controller := MemberResourcesCleanupController{
+		Client:    Client,
+		Scheme:    Scheme,
+		namespace: namespace,
+	}
+	return &controller
+}
+
+func (c *MemberResourcesCleanupController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	memberAnnounce := &mcv1alpha1.MemberClusterAnnounce{}
+	err := c.Get(ctx, req.NamespacedName, memberAnnounce)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	if err == nil && memberAnnounce.DeletionTimestamp.IsZero() {
+		// Try to clean up stale ResourceExports whenever there is a MemberClusterAnnounce
+		// creation event, to ensure that stale resources are removed in time so that
+		// there is no stale data for ResourceExport reconciler and being imported
```

Comment:
"so that they will not be processed by the ResourceExport reconciler and imported into member clusters"?